[13] Resource reconstruction failure case

Patch apktool by ourselves?

- [x] investigate
- [x] fix
- [x] deploy

```
I: Using Apktool 2.3.1-f7a872-SNAPSHOT
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether resources has changed...
I: Building resources...
W: invalid resource directory name: /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./res navigation
Exception in thread "main" brut.androlib.AndrolibException: brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [/var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/brut_util_Jar_6674335009107377142.tmp, p, --forced-package-id, 127, --min-sdk-version, 26, --target-sdk-version, 31, --version-code, 21122801, --version-name, 1.0.17, --no-version-vectors, -F, /var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/APKTOOL2172236797213094658.tmp, -0, arsc, -0, META-INF/androidx.activity_activity-ktx.version, -0, META-INF/androidx.activity_activity.version, -0, META-INF/androidx.annotation_annotation-experimental.version, -0, META-INF/androidx.appcompat_appcompat-resources.version, -0, META-INF/androidx.appcompat_appcompat.version, -0, META-INF/androidx.arch.core_core-runtime.version, -0, META-INF/androidx.asynclayoutinflater_asynclayoutinflater.version, -0, META-INF/androidx.browser_browser.version, -0, META-INF/androidx.camera_camera-camera2.version, -0, META-INF/androidx.camera_camera-core.version, -0, META-INF/androidx.camera_camera-lifecycle.version, -0, META-INF/androidx.camera_camera-view.version, -0, META-INF/androidx.cardview_cardview.version, -0, META-INF/androidx.coordinatorlayout_coordinatorlayout.version, -0, META-INF/androidx.core_core-ktx.version, -0, META-INF/androidx.core_core.version, -0, META-INF/androidx.cursoradapter_cursoradapter.version, -0, META-INF/androidx.customview_customview.version, -0, META-INF/androidx.databinding_viewbinding.version, -0, META-INF/androidx.documentfile_documentfile.version, -0, META-INF/androidx.drawerlayout_drawerlayout.version, -0, META-INF/androidx.dynamicanimation_dynamicanimation.version, -0, META-INF/androidx.emoji2_emoji2-views-helper.version, -0, META-INF/androidx.emoji2_emoji2.version, -0, META-INF/androidx.exifinterface_exifinterface.version, -0, META-INF/androidx.fragment_fragment-ktx.version, -0, META-INF/androidx.fragment_fragment.version, -0, META-INF/androidx.hilt_hilt-lifecycle-viewmodel.version, -0, META-INF/androidx.interpolator_interpolator.version, -0, META-INF/androidx.legacy_legacy-support-core-ui.version, -0, META-INF/androidx.legacy_legacy-support-core-utils.version, -0, META-INF/androidx.legacy_legacy-support-v4.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata.version, -0, META-INF/androidx.lifecycle_lifecycle-process.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-savedstate.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel.version, -0, META-INF/androidx.loader_loader.version, -0, META-INF/androidx.localbroadcastmanager_localbroadcastmanager.version, -0, META-INF/androidx.media_media.version, -0, META-INF/androidx.navigation_navigation-common-ktx.version, -0, META-INF/androidx.navigation_navigation-common.version, -0, META-INF/androidx.navigation_navigation-fragment-ktx.version, -0, META-INF/androidx.navigation_navigation-fragment.version, -0, META-INF/androidx.navigation_navigation-runtime-ktx.version, -0, META-INF/androidx.navigation_navigation-runtime.version, -0, META-INF/androidx.navigation_navigation-ui-ktx.version, -0, META-INF/androidx.navigation_navigation-ui.version, -0, META-INF/androidx.preference_preference-ktx.version, -0, META-INF/androidx.preference_preference.version, -0, META-INF/androidx.print_print.version, -0, META-INF/androidx.recyclerview_recyclerview.version, -0, META-INF/androidx.room_room-ktx.version, -0, META-INF/androidx.room_room-runtime.version, -0, META-INF/androidx.savedstate_savedstate-ktx.version, -0, META-INF/androidx.savedstate_savedstate.version, -0, META-INF/androidx.slidingpanelayout_slidingpanelayout.version, -0, META-INF/androidx.sqlite_sqlite-framework.version, -0, META-INF/androidx.sqlite_sqlite.version, -0, META-INF/androidx.startup_startup-runtime.version, -0, META-INF/androidx.swiperefreshlayout_swiperefreshlayout.version, -0, META-INF/androidx.tracing_tracing.version, -0, META-INF/androidx.transition_transition.version, -0, META-INF/androidx.vectordrawable_vectordrawable-animated.version, -0, META-INF/androidx.vectordrawable_vectordrawable.version, -0, META-INF/androidx.versionedparcelable_versionedparcelable.version, -0, META-INF/androidx.viewpager2_viewpager2.version, -0, META-INF/androidx.viewpager_viewpager.version, -0, META-INF/com.google.android.material_material.version, -0, META-INF/com.google.dagger_dagger-lint-aar.version, -0, META-INF/com.google.dagger_dagger.version, -0, META-INF/com.google.dagger_hilt-android.version, -0, META-INF/com.google.dagger_hilt-core.version, -0, arsc, -I, /Users/alterakey/Library/apktool/framework/1.apk, -S, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./res, -M, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./AndroidManifest.xml]
	at brut.androlib.Androlib.buildResourcesFull(Androlib.java:485)
	at brut.androlib.Androlib.buildResources(Androlib.java:419)
	at brut.androlib.Androlib.build(Androlib.java:318)
	at brut.androlib.Androlib.build(Androlib.java:270)
	at brut.apktool.Main.cmdBuild(Main.java:224)
	at brut.apktool.Main.main(Main.java:75)
Caused by: brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [/var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/brut_util_Jar_6674335009107377142.tmp, p, --forced-package-id, 127, --min-sdk-version, 26, --target-sdk-version, 31, --version-code, 21122801, --version-name, 1.0.17, --no-version-vectors, -F, /var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/APKTOOL2172236797213094658.tmp, -0, arsc, -0, META-INF/androidx.activity_activity-ktx.version, -0, META-INF/androidx.activity_activity.version, -0, META-INF/androidx.annotation_annotation-experimental.version, -0, META-INF/androidx.appcompat_appcompat-resources.version, -0, META-INF/androidx.appcompat_appcompat.version, -0, META-INF/androidx.arch.core_core-runtime.version, -0, META-INF/androidx.asynclayoutinflater_asynclayoutinflater.version, -0, META-INF/androidx.browser_browser.version, -0, META-INF/androidx.camera_camera-camera2.version, -0, META-INF/androidx.camera_camera-core.version, -0, META-INF/androidx.camera_camera-lifecycle.version, -0, META-INF/androidx.camera_camera-view.version, -0, META-INF/androidx.cardview_cardview.version, -0, META-INF/androidx.coordinatorlayout_coordinatorlayout.version, -0, META-INF/androidx.core_core-ktx.version, -0, META-INF/androidx.core_core.version, -0, META-INF/androidx.cursoradapter_cursoradapter.version, -0, META-INF/androidx.customview_customview.version, -0, META-INF/androidx.databinding_viewbinding.version, -0, META-INF/androidx.documentfile_documentfile.version, -0, META-INF/androidx.drawerlayout_drawerlayout.version, -0, META-INF/androidx.dynamicanimation_dynamicanimation.version, -0, META-INF/androidx.emoji2_emoji2-views-helper.version, -0, META-INF/androidx.emoji2_emoji2.version, -0, META-INF/androidx.exifinterface_exifinterface.version, -0, META-INF/androidx.fragment_fragment-ktx.version, -0, META-INF/androidx.fragment_fragment.version, -0, META-INF/androidx.hilt_hilt-lifecycle-viewmodel.version, -0, META-INF/androidx.interpolator_interpolator.version, -0, META-INF/androidx.legacy_legacy-support-core-ui.version, -0, META-INF/androidx.legacy_legacy-support-core-utils.version, -0, META-INF/androidx.legacy_legacy-support-v4.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata.version, -0, META-INF/androidx.lifecycle_lifecycle-process.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-savedstate.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel.version, -0, META-INF/androidx.loader_loader.version, -0, META-INF/androidx.localbroadcastmanager_localbroadcastmanager.version, -0, META-INF/androidx.media_media.version, -0, META-INF/androidx.navigation_navigation-common-ktx.version, -0, META-INF/androidx.navigation_navigation-common.version, -0, META-INF/androidx.navigation_navigation-fragment-ktx.version, -0, META-INF/androidx.navigation_navigation-fragment.version, -0, META-INF/androidx.navigation_navigation-runtime-ktx.version, -0, META-INF/androidx.navigation_navigation-runtime.version, -0, META-INF/androidx.navigation_navigation-ui-ktx.version, -0, META-INF/androidx.navigation_navigation-ui.version, -0, META-INF/androidx.preference_preference-ktx.version, -0, META-INF/androidx.preference_preference.version, -0, META-INF/androidx.print_print.version, -0, META-INF/androidx.recyclerview_recyclerview.version, -0, META-INF/androidx.room_room-ktx.version, -0, META-INF/androidx.room_room-runtime.version, -0, META-INF/androidx.savedstate_savedstate-ktx.version, -0, META-INF/androidx.savedstate_savedstate.version, -0, META-INF/androidx.slidingpanelayout_slidingpanelayout.version, -0, META-INF/androidx.sqlite_sqlite-framework.version, -0, META-INF/androidx.sqlite_sqlite.version, -0, META-INF/androidx.startup_startup-runtime.version, -0, META-INF/androidx.swiperefreshlayout_swiperefreshlayout.version, -0, META-INF/androidx.tracing_tracing.version, -0, META-INF/androidx.transition_transition.version, -0, META-INF/androidx.vectordrawable_vectordrawable-animated.version, -0, META-INF/androidx.vectordrawable_vectordrawable.version, -0, META-INF/androidx.versionedparcelable_versionedparcelable.version, -0, META-INF/androidx.viewpager2_viewpager2.version, -0, META-INF/androidx.viewpager_viewpager.version, -0, META-INF/com.google.android.material_material.version, -0, META-INF/com.google.dagger_dagger-lint-aar.version, -0, META-INF/com.google.dagger_dagger.version, -0, META-INF/com.google.dagger_hilt-android.version, -0, META-INF/com.google.dagger_hilt-core.version, -0, arsc, -I, /Users/alterakey/Library/apktool/framework/1.apk, -S, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./res, -M, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./AndroidManifest.xml]
	at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:456)
	at brut.androlib.Androlib.buildResourcesFull(Androlib.java:471)
	... 5 more
Caused by: brut.common.BrutException: could not exec (exit code = 1): [/var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/brut_util_Jar_6674335009107377142.tmp, p, --forced-package-id, 127, --min-sdk-version, 26, --target-sdk-version, 31, --version-code, 21122801, --version-name, 1.0.17, --no-version-vectors, -F, /var/folders/y7/cvybpmkn65qbzrh3hygvhknw0000gn/T/APKTOOL2172236797213094658.tmp, -0, arsc, -0, META-INF/androidx.activity_activity-ktx.version, -0, META-INF/androidx.activity_activity.version, -0, META-INF/androidx.annotation_annotation-experimental.version, -0, META-INF/androidx.appcompat_appcompat-resources.version, -0, META-INF/androidx.appcompat_appcompat.version, -0, META-INF/androidx.arch.core_core-runtime.version, -0, META-INF/androidx.asynclayoutinflater_asynclayoutinflater.version, -0, META-INF/androidx.browser_browser.version, -0, META-INF/androidx.camera_camera-camera2.version, -0, META-INF/androidx.camera_camera-core.version, -0, META-INF/androidx.camera_camera-lifecycle.version, -0, META-INF/androidx.camera_camera-view.version, -0, META-INF/androidx.cardview_cardview.version, -0, META-INF/androidx.coordinatorlayout_coordinatorlayout.version, -0, META-INF/androidx.core_core-ktx.version, -0, META-INF/androidx.core_core.version, -0, META-INF/androidx.cursoradapter_cursoradapter.version, -0, META-INF/androidx.customview_customview.version, -0, META-INF/androidx.databinding_viewbinding.version, -0, META-INF/androidx.documentfile_documentfile.version, -0, META-INF/androidx.drawerlayout_drawerlayout.version, -0, META-INF/androidx.dynamicanimation_dynamicanimation.version, -0, META-INF/androidx.emoji2_emoji2-views-helper.version, -0, META-INF/androidx.emoji2_emoji2.version, -0, META-INF/androidx.exifinterface_exifinterface.version, -0, META-INF/androidx.fragment_fragment-ktx.version, -0, META-INF/androidx.fragment_fragment.version, -0, META-INF/androidx.hilt_hilt-lifecycle-viewmodel.version, -0, META-INF/androidx.interpolator_interpolator.version, -0, META-INF/androidx.legacy_legacy-support-core-ui.version, -0, META-INF/androidx.legacy_legacy-support-core-utils.version, -0, META-INF/androidx.legacy_legacy-support-v4.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-core.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-livedata.version, -0, META-INF/androidx.lifecycle_lifecycle-process.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-runtime.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-ktx.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel-savedstate.version, -0, META-INF/androidx.lifecycle_lifecycle-viewmodel.version, -0, META-INF/androidx.loader_loader.version, -0, META-INF/androidx.localbroadcastmanager_localbroadcastmanager.version, -0, META-INF/androidx.media_media.version, -0, META-INF/androidx.navigation_navigation-common-ktx.version, -0, META-INF/androidx.navigation_navigation-common.version, -0, META-INF/androidx.navigation_navigation-fragment-ktx.version, -0, META-INF/androidx.navigation_navigation-fragment.version, -0, META-INF/androidx.navigation_navigation-runtime-ktx.version, -0, META-INF/androidx.navigation_navigation-runtime.version, -0, META-INF/androidx.navigation_navigation-ui-ktx.version, -0, META-INF/androidx.navigation_navigation-ui.version, -0, META-INF/androidx.preference_preference-ktx.version, -0, META-INF/androidx.preference_preference.version, -0, META-INF/androidx.print_print.version, -0, META-INF/androidx.recyclerview_recyclerview.version, -0, META-INF/androidx.room_room-ktx.version, -0, META-INF/androidx.room_room-runtime.version, -0, META-INF/androidx.savedstate_savedstate-ktx.version, -0, META-INF/androidx.savedstate_savedstate.version, -0, META-INF/androidx.slidingpanelayout_slidingpanelayout.version, -0, META-INF/androidx.sqlite_sqlite-framework.version, -0, META-INF/androidx.sqlite_sqlite.version, -0, META-INF/androidx.startup_startup-runtime.version, -0, META-INF/androidx.swiperefreshlayout_swiperefreshlayout.version, -0, META-INF/androidx.tracing_tracing.version, -0, META-INF/androidx.transition_transition.version, -0, META-INF/androidx.vectordrawable_vectordrawable-animated.version, -0, META-INF/androidx.vectordrawable_vectordrawable.version, -0, META-INF/androidx.versionedparcelable_versionedparcelable.version, -0, META-INF/androidx.viewpager2_viewpager2.version, -0, META-INF/androidx.viewpager_viewpager.version, -0, META-INF/com.google.android.material_material.version, -0, META-INF/com.google.dagger_dagger-lint-aar.version, -0, META-INF/com.google.dagger_dagger.version, -0, META-INF/com.google.dagger_hilt-android.version, -0, META-INF/com.google.dagger_hilt-core.version, -0, arsc, -I, /Users/alterakey/Library/apktool/framework/1.apk, -S, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./res, -M, /Users/alterakey/.trueseeing2/84/67/c28025d6bbc86ff7c0280169ed3678f61dc2673f360984a09de5cccd702b/./AndroidManifest.xml]
	at brut.util.OS.exec(OS.java:95)
	at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:450)
	... 6 more
```

--- a/trueseeing/core/patch.py
+++ b/trueseeing/core/patch.py
@@ -46,7 +46,7 @@ class Patcher:
       context.analyze()
       ui.info(f"{self.apk} -> {context.wd}")
       for p in patches:
-          p.apply(context)
+        p.apply(context)
 
       # XXX
       sigfile = 'CERT'

--- a/trueseeing/core/patch.py
+++ b/trueseeing/core/patch.py
@@ -55,6 +55,6 @@ class Patcher:
       with TemporaryDirectory() as d:
         from trueseeing.core.tools import invoke_passthru
         invoke_passthru("(mkdir -p {root}/)".format(root=d))
-        invoke_passthru("(cd {wd} && java -jar {apktool} b -o {root}/patched.apk .)".format(root=d, apktool=resource_filename(__name__, os.path.join('..', 'libs', 'apktool.jar')), wd=context.wd))
+        invoke_passthru("(cd {wd} && java -jar {apktool} b --use-aapt2 -o {root}/patched.apk .)".format(root=d, apktool=resource_filename(__name__, os.path.join('..', 'libs', 'apktool.jar')), wd=context.wd))
         invoke_passthru("(cd {root} && jarsigner -sigalg SHA1withRSA -digestalg SHA1 -keystore {keystore} -storepass android -keypass android -sigfile {sigfile} patched.apk androiddebugkey)".format(root=d, keystore=SigningKey().key(), sigfile=sigfile))
         copyfile(os.path.join(d, 'patched.apk'), self.out)

--- a/trueseeing/signature/security.py
+++ b/trueseeing/signature/security.py
@@ -304,13 +304,13 @@ class SecurityInsecureWebViewDetector(Detector):
                         source=store.query().qualname_of(q)
                       )
                   except (DataFlows.NoSuchValueError):
-                      yield Issue(
-                        detector_id=self.option,
-                        confidence='tentative',
-                        cvss3_vector=self._cvss,
-                        summary=self._summary1,
-                        source=store.query().qualname_of(q)
-                      )
+                    yield Issue(
+                      detector_id=self.option,
+                      confidence='tentative',
+                      cvss3_vector=self._cvss,
+                      summary=self._summary1,
+                      source=store.query().qualname_of(q)
+                    )
           except (DataFlows.NoSuchValueError):
             pass
 
@@ -414,11 +414,11 @@ class LogDetector(Detector):
               source=store.query().qualname_of(cl)
             )
         else:
-            yield Issue(
-              detector_id=self.option,
-              confidence='tentative',
-              cvss3_vector=self._cvss,
-              summary=self._summary,
-              info1=cl.p[1].v,
-              source=store.query().qualname_of(cl)
-            )
+          yield Issue(
+            detector_id=self.option,
+            confidence='tentative',
+            cvss3_vector=self._cvss,
+            summary=self._summary,
+            info1=cl.p[1].v,
+            source=store.query().qualname_of(cl)
+          )

